src/Sulu/Bundle/MediaBundle/Trash/MediaTrashItemHandler.php

```diff
@@ -0,0 +1,344 @@
+use Sulu\Bundle\MediaBundle\Entity\MediaType;
+use Sulu\Bundle\TagBundle\Tag\TagInterface;
+use Sulu\Bundle\TrashBundle\Application\DoctrineRestoreHelper\DoctrineRestoreHelperInterface;
+use Sulu\Bundle\TrashBundle\Application\TrashItemHandler\RestoreTrashItemHandlerInterface;
+use Sulu\Bundle\TrashBundle\Application\TrashItemHandler\StoreTrashItemHandlerInterface;
+use Sulu\Bundle\TrashBundle\Domain\Model\TrashItemInterface;
+use Sulu\Bundle\TrashBundle\Domain\Repository\TrashItemRepositoryInterface;
+use Sulu\Component\Security\Authentication\UserInterface;
+use Webmozart\Assert\Assert;
+
+final class MediaTrashItemHandler implements StoreTrashItemHandlerInterface, RestoreTrashItemHandlerInterface
+{
+    /**
+     * @var TrashItemRepositoryInterface
+     */
+    private $trashItemRepository;
+
+    /**
+     * @var MediaRepositoryInterface
+     */
+    private $mediaRepository;
+
+    /**
+     * @var EntityManagerInterface
+     */
+    private $entityManager;
+
+    /**
+     * @var DoctrineRestoreHelperInterface
+     */
+    private $doctrineRestoreHelper;
+
+    /**
+     * @var DomainEventCollectorInterface
+     */
+    private $domainEventCollector;
+
+    public function __construct(
+        TrashItemRepositoryInterface $trashItemRepository,
+        MediaRepositoryInterface $mediaRepository,
+        EntityManagerInterface $entityManager,
+        DoctrineRestoreHelperInterface $doctrineRestoreHelper,
+        DomainEventCollectorInterface $domainEventCollector
+    ) {
+        $this->trashItemRepository = $trashItemRepository;
+        $this->mediaRepository = $mediaRepository;
+        $this->entityManager = $entityManager;
+        $this->doctrineRestoreHelper = $doctrineRestoreHelper;
+        $this->domainEventCollector = $domainEventCollector;
+    }
+
+    /**
+     * @param MediaInterface $media
+     */
+    public function store(object $media): TrashItemInterface
+    {
+        Assert::isInstanceOf($media, MediaInterface::class);
+
+        // TODO: move original image file
+
+        $creator = $media->getCreator();
+        $previewImage = $media->getPreviewImage();
+
+        $data = [
+            'typeId' => $media->getType()->getId(),
+            'previewImageId' => $previewImage ? $previewImage->getId() : null,
+            'created' => $media->getCreated()->format('c'),
+            'creatorId' => $creator ? $creator->getId() : null,
+            'files' => [],
+        ];
+        $mediaTitles = [];
+
+        /** @var File $file */
+        foreach ($media->getFiles() as $file) {
+            $creator = $file->getCreator();
+
+            $fileData = [
+                'version' => $file->getVersion(),
+                'created' => $file->getCreated()->format('c'),
+                'creatorId' => $creator ? $creator->getId() : null,
+                'fileVersions' => [],
+            ];
+
+            /** @var FileVersion $fileVersion */
+            foreach ($file->getFileVersions() as $fileVersion) {
```

Comment:
/cc @chirimoya should we move all versions into the trash?

Reply:
we decided to store/move all versions for now 🙂